README.md
![balenaDash](https://www.balena.io/blog/content/images/2019/03/balenaDash_logo.png)

Try our simple [balenaCloud](https://www.balena.io/cloud) starter project to set up a Raspberry Pi and LCD or monitor to display any webpage or web-based dashboard. We’re creatively calling it balenaDash! Follow along with our guide to get up and running in under 30 minutes.

* [Introduction](#introduction)
* [Hardware required](#hardware-required)
* [Using WiFi Connect](#using-wifi-connect)
* [Controlling content](#controlling-content)
* [Automate backlight switching](#automate-backlight-switching)
* [Photo slideshow feature](#photo-slideshow-feature)

## Introduction

This is the perfect introductory project to try out [balenaCloud](https://www.balena.io/cloud) and see how everything works.

At the most basic level, this project allows you to display any webpage using a lightweight web browser. This means that you can build a device dedicated to showing anything that runs in a normal web browser, it will boot up and automatically start displaying what you choose.

Some examples of what you could use this for include:

* [Instagram](https://instagram.com) photo stream (based on hashtag or user)
* Data-centric dashboards using [Grafana](https://grafana.com/) or [Datadog](https://www.datadoghq.com/)
* Live digital photo frame feeding from [Google Photos](https://photos.google.com/)
* Display for services such as [Flightradar24](https://www.flightradar24.com/) or  [Flightaware](https://flightaware.com/)
* Digital [fishtank](http://www.fishgl.com/) or [jellyfish](https://arodic.github.io/p/jellyfish/)
* Streaming webcam display
* 24 hour live [animated cat](http://www.nyan.cat/) display
* [Home automation](https://www.home-assistant.io) dashboard

## Hardware required

We’ve partnered with our friends at [PiSupply](https://pisupp.ly/) to make it easy for you to get ahold of all the hardware you need to build this project in a single package. We’ve created two packages, one that includes everything you’ll need, and a second that includes everything apart from the display for if you’ve already got a monitor you want to use.

* **[balenaDash kit](https://uk.pi-supply.com/products/balenadash-kit)**
* **[balenaDash kit - barebones](https://uk.pi-supply.com/products/balenadash-kit-barebones)**

The list of items you’ll need is also included below:

* Raspberry Pi 3B/3B+ (**Note:** this project will not work with the Pi Zero or older devices with < 1GB RAM)
* 16GB Micro-SD Card (we recommend Sandisk Extreme Pro SD cards)
* Display (any Raspberry Pi display will work for this project)
* Micro-USB cable
* Power supply
* Case (optional)

![](https://www.balena.io/blog/content/images/2018/11/image17.jpg)

## Setup and configuration

Use of balenaDash requires that you allocate more memory to the GPU. This is achieved by adding (or editing the existing) the **Device configuration variable** `RESIN_HOST_CONFIG_gpu_mem`, for this project we recommend setting it to `396`.

### Official Raspberry Pi 7-inch display

If you are using the official [Raspberry Pi 7 inch display](https://www.raspberrypi.org/products/raspberry-pi-touch-display/), you can follow [this tutorial](https://www.balena.io/blog/assembling-the-official-raspberry-pi-touchscreen) to assemble and configure the screen.

Depending on the orientation of the majority of your content or photos, you can choose to have the display in horizontal or vertical mode. On **Fleet Configuration** add a variable called `BALENA_HOST_CONFIG_display_lcd_rotate` with value `2` for horizontal (180º rotation) or `1` for vertical (90º clockwise rotation). More details about the options for this are available [on the Raspberry Pi site](https://www.raspberrypi.org/documentation/configuration/config-txt/video.md).

## Using WiFi Connect

The balenaDash project includes [wifi-connect](https://github.com/balena-io/wifi-connect) which enables your device to operate as a WiFi access point and allow you to join a different WiFi network using captive portal functionality. Although you can specify a WiFi network to join when you first add your device and download the image from the balenaCloud dashboard, there may be situations where you need to change that.

WiFi Connect periodically tests for a functional internet connection, if nothing is found the device sets itself up as a WiFi access point named `balenaDash` that you can join with a mobile device.

To use WiFi Connect you need to join the `balenaDash` network and you should see a captive portal popup. The passphrase is `balenaDash`. If not, ensure that you remain connected to the `balenaDash` network and visit the IP address of the device in a browser on port `80`. For example `http://<ip of balenaDash device>`. This will allow you to access WiFi Connect, perform a site survey and join a different WiFi network.

## Controlling content

### Loading a URL

To configure the URL displayed by webkit, set the **`WPE_URL`** environment
variable. If nothing is set, balenaDash will display the balenaOS logo on the screen.

### Switching URLs quickly using your web browser or Slack

Your balenaDash device is also running a small webserver on port 8080. The screen will show the URL configured at `WPE_URL` normally, but the webserver allows you to put other URLs on screen quickly and easily. If you tell balenaCloud to expose your device's public URL, then you can even control it with Slack (or `curl`, or anything that can use webhooks).  [More details](https://github.com/mozz100/tohora/blob/master/README.md)

## Automate backlight switching
To use automatic backlight switching you’ll need to configure a few service variables for the scheduler service.

`ENABLE_BACKLIGHT_TIMER=1`
`BACKLIGHT_ON=0 8 * * *`
`BACKLIGHT_OFF=0 23 * * *`

The `BACKLIGHT_ON` and `BACKLIGHT_OFF` variables accept standard cron syntax; take a look at https://crontab.guru if you’re not familiar. For more instructions check out [our blog post](https://www.balena.io/blog/automate-the-backlight-timer-on-your-balenadash-display/).

## Photo slideshow feature

One extra feature implemented on balenaDash is the ability to use it to display a photo slideshow. Out of the box you can use [Google Photo Albums](https://photos.google.com), [Dropbox Photo Album](https://www.dropbox.com) and [Apple Photos](https://www.icloud.com/). The images are downloaded automatically and auto updated in case of changes.

The main objective of this project is to create a physical photo album that is easy to build, configure, and maintain. Once configured, there is nothing else you need to worry about. If you edit the album, add new or delete photos, as soon as the device restarts it will be updated with the last changes.

This is the perfect gift for family and friends. Give a smart photo album to your mother or grandmother so that they can keep up with your baby pictures, for example.

#### Google Photos Albums

![](https://raw.githubusercontent.com/balena-io-playground/balena-dash-google-photos/master/media/google_photos.gif)

Go to [https://photos.google.com](https://photos.google.com) and select the album that you want to share.

Click in the share button, click in the `Create link` button and copy the the url.
Example: **https://photos.app.goo.gl/rAnDoMvAlUe123**.

#### Dropbox Photo Albums

![](https://raw.githubusercontent.com/balena-io-playground/balena-dash-google-photos/master/media/dropbox.gif)

On the [dropbox](https://www.dropbox.com/home) website, go to the folder that contains the photos and click on `Share Folder` and then `Copy link`. This will the URL you will need to add to balenaCloud.

#### Apple iCloud Photo Album

Create a photo album and copy the share url, similar to `https://www.icloud.com/sharedalbum/#ALBUM-ID`

#### USB drive Photo Album

Place your photos on a USB stick and plug it into the Raspberry Pi. Each time all previous existing photos will be removed and replaced by the new ones. 
- Set ```GALLERY_URL``` to ```USBDRIVE``` to use this mode.
- Make sure to update ```CRON_SCHEDULE``` accordingly or else image changes will only be picked up at reboot.


#### Setting up the photo album

* On balenaCloud, go to **Device variables D(x)** and add the following:

| ENV VAR                 | Description                                                                                                                                           | Options                                        | Default       |
|-------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------|---------------|
| GALLERY_URL             | Gallery URL for **google photos**, **dropbox images**, **apple photos** or **usb drive**.                                                                             |                                                |               |
| GALLERY_SLIDESHOW_DELAY | Slideshow delay in milliseconds                                                                                                                       |                                                | 10000         |
| GALLERY_IMAGE_STYLE     | `Contain` shows the entire image on the screen. `Cover` zooms the image filling the entire screen.                                                    | contain,  cover                                | cover         |
| GALLERY_EFFECT          | Transition effects                                                                                                                                    | fade,  horizontal,  vertical,  kenburns, false | fade          |
| CRON_SCHEDULE           | Cron scheduler to reload images to get changes                                                                                                        |                                                | 0 */12 * * *  |
| SHUFFLE_SLIDESHOW       | Shiffle images to display randomly                                                                                                                    | true, false                                    | false         |
| RESIZE_WIDTH            | * Resize image width or height (larger side) in pixels                                                                                                |                                                | 1000px        |
| COMPRESS_QUALITY        | * Image compression                                                                                                                                   | 0 - 100                                        | 90            |


    * Only available for iCloud photos

Note that after some performance tests on the Raspberry Pi 2 & 3, the combination of `GALLERY_IMAGE_STYLE = contain` and `GALLERY_EFFECT = fade or kenburns` can make the transition effects choppy.

### For a complete tutorial on how to use balenaDash, please check out our blog post at [https://www.balena.io/blog/make-a-web-frame-with-raspberry-pi-in-30-minutes/](https://www.balena.io/blog/make-a-web-frame-with-raspberry-pi-in-30-minutes/)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.8.0
## (2019-10-23)

* Moved sleep in wpe-init for custom url support [Anže Jenšterle]
* var check for scheduler script [Anže Jenšterle]
* Added USB stick support for photos container. [Tomás Migone]
* Update readme file [Daniel Andrade]
* Add possibility to shuffle images. Closes #33 [Daniel Andrade]

# v0.7.4
## (2019-10-09)

* Add FAQ [Dimitrios Lytras]

# v0.7.3
## (2019-10-08)

* Update codeowners, add PR template [Chris Crocker-White]

# v0.7.2
## (2019-10-01)

* photos: switch to multistage build [Shaun Mulligan]

# v0.7.1
## (2019-09-30)

* add sleep when starting wifi-connect [Shaun Mulligan]

# v0.7.0
## (2019-09-25)

* Add machine-readable changelog [Dimitrios Lytras]
* Enable scrutinizer [Dimitrios Lytras]

# v0.6.1
## (2019-08-19)

* Moved sleep in wpe-init for custom url support [Anže Jenšterle]
* var check for scheduler script [Anže Jenšterle]

# v0.6.0
## (2019-08-05)

* Update readme file [Daniel Andrade]
* Add possibility to shuffle images. Closes #33 [Daniel Andrade]

# v0.5.2
## (2019-07-31)

* Add contribution.md [Chris Crocker-White]

# v0.5.1
## (2019-07-26)

* Fix readme file to remove youtube-tv as default. [Daniel Andrade]

# v0.5.0
## (2019-07-26)

* Fix issue where is wasnt showing youtube if no variables are set [Daniel Andrade]

# v0.4.1
## (2019-07-25)

* Update README with gpu_mem and display_lcd_rotate info [Chris Crocker-White]

# v0.4.0
## (2019-07-23)

* Add photo slideshow option to balena dash [Daniel Andrade]

# v0.3.0
## (2019-06-13)

* Testing with marhoy's image [Chris Crocker-White]
* Prevent multiple cog instances - use only Tohora [Chris Crocker-White]
* Bump balena-wpe to latest version and update init [Chris Crocker-White]

# v0.2.0
## (2019-04-09)

* Remove wifi-connect note [Daniel Andrade]
* Update readme file to change wifi-connect to port 80 and web-ui to 8080 [Daniel Andrade]
* Set wifi-connect to port 80 and web-ui to port 8080 [Daniel Andrade]

# v0.1.1
## (2019-04-03)

* Make WiFi connect heading match ToC [Chris Crocker-White]
* Add info for wifi-connect to readme [Chris Crocker-White]

# v0.1.0
## (2019-04-01)

* Update comments in wifi-connect [Chris Crocker-White]
* Add CODEOWNERS [Chris Crocker-White]
* Added wifi-connect on port 8080 [Chris Crocker-White]
* Modify readme to put new Tohora functionality in correct place, add repo.yml for versionist [Chris Crocker-White]
* Changing to multicontainer + adding scheduler [Chris Crocker-White]
* fix: fix up file permisions [Shaun Mulligan]


photos/server.js
// Photo Album Server Setup
// =============================================================================

// Call the packages we need
var express   = require("express");
var request   = require("request");
var rp        = require("request-promise-native");
var cors      = require("cors");
var _chunk    = require("lodash.chunk");
var fs        = require("fs");
var path      = require("path");
var sharp     = require("sharp");
var schedule  = require("node-schedule");

var app       = express();

// Only run server if GALLERY_URL is defined
if(process.env.GALLERY_URL) {

  // Enable HTML template middleware
  app.engine("html", require("ejs").renderFile);

  // Enable static CSS styles
  app.use(express.static("styles"));

  // Set server port
  var port = process.env.PORT || 8888;

  // Get an instance of the express router
  var router = express.Router();

  // Main route
  router.get("/", cors(), function(req, res) {

    // Wait images to be processed
    if(images.length > 0){
      res.render("index.html", {
        loadingCompleted: true,
        photos: images,
        timer: slideshow_timer,
        image_styles: image_styles,
        transition: transition
      });

    } else {
      res.render("index.html", {
        loadingCompleted: false,
        photos: [],
        timer: slideshow_timer,
        image_styles: image_styles,
        transition: transition
      });
    }

  });

  // Get the next image in the array
  router.get("/img/:id", function(req, res) {

    var next = {};
    if (req.params.id < images.length ) {
      // return next image on the list
      next.img = images[req.params.id++];
      next.id = req.params.id++;
    } else {
      // return first image
      next.img = images[0];
      next.id = 0;
    }

    res.send(next);

  });

  // Register routes
  app.use(express.static(__dirname));
  app.use("/", router);

  // Get image variables
  var images = [];
  fetchImages(process.env.GALLERY_URL);

  // Set slideshow time
  var slideshow_timer = 10000; // default value
  if (process.env.GALLERY_SLIDESHOW_DELAY) {
    var slideshow_timer = parseInt(process.env.GALLERY_SLIDESHOW_DELAY);
  }

  // Set photo scaling
  var image_styles = "cover"; // default value
  if (process.env.GALLERY_IMAGE_STYLE) {
    var styles = ["contain", "cover"];
    var container = process.env.GALLERY_IMAGE_STYLE.toLowerCase();
    if (styles.includes(container)) {
      image_styles = container;
    }
  } 

  // Gallery transition effect
  var transition = "fade" ; // default value
  if (process.env.GALLERY_EFFECT) {
    var effects = ["fade", "horizontal", "vertical", "kenburns", "false"];
    var transitionSelected = process.env.GALLERY_EFFECT.toLowerCase();
    if (effects.includes(transitionSelected)) {
      transition = transitionSelected; 
    }
  } 

  // Get gallery update cron timing
  // 0 */12 * * *
  var cronSchedule = "0 */12 * * *"; // default value
  if (process.env.CRON_SCHEDULE) {
    cronSchedule = process.env.CRON_SCHEDULE;
    console.log("🐉 - Scheduler configured for "+cronSchedule);
  }

  // Get resize width
  var maxWidth = 1000; // default value
  if (process.env.RESIZE_WIDTH) {
    maxWidth = parseInt(process.env.RESIZE_WIDTH);
  }

  // Get image compress quality
  var compressQuality = 90; // default value
  if (process.env.COMPRESS_QUALITY) {
    compressQuality = parseInt(process.env.COMPRESS_QUALITY);
  }
  
  // Shuffle album images
  var shuffleGallery = false; // default value
  if (process.env.SHUFFLE_SLIDESHOW) {
    var shuffleGallery = (process.env.SHUFFLE_SLIDESHOW == "true");
  }

  // Server Start
  // =============================================================================
  app.listen(port);
  console.log("✨ - Magic happens on port " + port);

  // Update image scheduler
  var j = schedule.scheduleJob(cronSchedule, function(){
    images = [];
    console.log("🐉 - Refreshing gallery array");
    fetchImages(process.env.GALLERY_URL);
  });


} else {
  app.get('/', (req, res) => res.send('Hello there!!'))
  app.listen(port);
  console.log("GALLERY_URL not set, stopping express server.");
}

// Functions //

/**
 * Fetch all images from a custom album url 
 * @param {String} albumURL Album URL from Google Photos, Dropbox or Apple Photos.
 */
function fetchImages(albumURL) {
  console.log("📷 - Fetching images...");

  var hostname = getHostName(albumURL);

  // Google Photos
  if (["photos.app.goo.gl", "goo.gl"].includes(hostname)) {
    console.log("📷 - Starting a google photos slideshow.");

    var photos = [];

    // Parse photos
    const rx = /\["(https:\/\/[^\.]+.googleusercontent\.com\/[^"]+)",([0-9]+),([0-9]+),/;
    const extractPhotos = data =>
      data
        .match(new RegExp(rx, "g"))
        .map(m => m.match(rx))
        .map(p => {
          const width = +p[2];
          const height = +p[3];
          const url = `${p[1]}=w${width}-h${height}-no`;
          return `${p[1]}=w1000`; // Change here to increase/decrease image width-size
        });

    request(
      {
        uri: albumURL,
        method: "GET",
        timeout: 10000
      },
      function(error, response, body) {
        if (!error && response.statusCode == 200) {
          // Fetch and remove duplicate images
          photos = extractPhotos(body);
          photos = [...new Set(photos)]; // save only unique values

          console.log("Found " + photos.length + " images.");
          setImagesArray(photos);
        }

        // In case of errors
        else {
          photos.push("views/album_url_error.png");
          setImagesArray(photos);
          console.log("error:", error);
        }
      }
    );
  }

  // Dropbox Photos
  else if (hostname == "dropbox.com") {
    console.log("📷 - Starting a new dropbox photos slideshow.");

    var photos = [];

    const rx = /(https:\/\/)([^\.]+)(.previews.dropboxusercontent.com\/p\/thumb\/)([^\/]+)(\/p.jpeg 1280)/;
    const extractPhotos = data => data.match(new RegExp(rx, "g"));

    request(
      {
        uri: albumURL,
        method: "GET",
        timeout: 10000
      },
      function(error, response, body) {
        if (!error && response.statusCode == 200) {
          body = extractPhotos(body);

          body.forEach(function(entry) {
            entry = entry.slice(0, -5);
            photos.push(entry);
          });

          console.log("Found " + photos.length + " images.");
          setImagesArray(photos);
        }

        // In case of errors
        else {
          photos.push("views/album_url_error.png");
          setImagesArray(photos);
          console.log("error:", error);
        }
      }
    );
  }

  // Apple Photos
  else if (hostname == "icloud.com") {
    console.log("📷 - Starting a new apple photos slideshow.");

    // Get the album id from URL
    var albumId = albumURL.match(new RegExp("([^#/]+$)", "g"))[0];

    // Calculate base URL
    var baseUrl = getBaseUrl(albumId);

    // Download images
    getPhotoMetadata(baseUrl)
      .then(function(metadata) {
        var chunks = _chunk(metadata.photoGuids, 25);

        var processChunks = function(i) {
          if (i < chunks.length) {
            getUrls(baseUrl, chunks[i]).then(function(urls) {
              decorateUrls(metadata, urls);

              setTimeout(function() {
                processChunks(i + 1);
              }, 1000);
            });
          } else {
            var i = 0;
            for (var photoGuid in metadata.photos) {
              var photo = metadata.photos[photoGuid];

              // Download images
              console.log("[file"+i+".jpg] - Downloading");
              downloadFile(photo.url, "file" + i + ".jpg");
              i++;
            }

            images = listImagesInDir(path.dirname(require.main.filename));
          }
        };

        processChunks(0);
      })
      .catch(function(error) {
        images.push("views/album_url_error.png");
        setImagesArray(images);
        console.log("error:", error);
      });
  } 
  else if (albumURL === "USBDRIVE") {
    let usbPhotos = listImagesInDirSync('/usbstorage')
    let currentPhotos = listImagesInDirSync(path.join(path.dirname(require.main.filename), 'views/usbstorage'))
    
    try {
      if (usbPhotos.length > 0) {
        // Only update photos if there are new ones on /usbstorage
        currentPhotos.map(p => fs.unlinkSync(p))
        usbPhotos.map(p => fs.copyFileSync(p, `views/usbstorage/${path.basename(p)}`))
        let newPhotos = listImagesInDirSync('views/usbstorage')
        setImagesArray(newPhotos);
        console.log(`Got ${newPhotos.length} image(s) from /usbstorage.`);
      } else if (currentPhotos.length === 0) {
        images.push("views/album_url_error.png");
        setImagesArray(images);
        console.log("No images found on views/usbstorage folder.");
      }
    } catch (error) {
      images.push("views/album_url_error.png");
      setImagesArray(images);
      console.log("error:", error);
    }
  }
  else {
    images.push("views/album_url_error.png");
    setImagesArray(images);
    console.log("error: Could not parse album.");
  }
}

/**
 * Callback function to set image array.
 * Depending on the parameters, can also suffle image array. 
 * @param {Array} value Images array.
 */
function setImagesArray(value) {
  // callback function
  if(shuffleGallery) {
    console.log("🔀 - Shuffling images ON 🔀")
    images = shuffle(value);
  } else {
    console.log("🔀 - Shuffling images OFF 🔀")
    images = value;
  }
}

/**
 * From a given URL, gets the hostname
 * @param {String} url Website url.
 */
function getHostName(url) {
  var match = url.match(/:\/\/(www[0-9]?\.)?(.[^/:]+)/i);
  if (
    match != null &&
    match.length > 2 &&
    typeof match[2] === "string" &&
    match[2].length > 0
  ) {
    return match[2];
  } else {
    return null;
  }
}

/**
 * Parse the base url for apple photos
 * @param {String} token Token is the album id.
 */
function getBaseUrl(token) {
  var BASE_62_CHAR_SET =
    "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

  var base62ToInt = function(e) {
    var t = 0;
    for (var n = 0; n < e.length; n++)
      t = t * 62 + BASE_62_CHAR_SET.indexOf(e[n]);
    return t;
  };

  var e = token,
    t = e[0],
    n = t === "A" ? base62ToInt(e[1]) : base62ToInt(e.substring(1, 3)),
    r = e,
    i = e.indexOf(";"),
    s = null;

  if (i >= 0) {
    s = e.slice(i + 1);
    r = r.replace(";" + s, "");
  }

  var serverPartition = n;

  var baseUrl = "https://p";

  baseUrl += serverPartition < 10 ? "0" + serverPartition : serverPartition;
  baseUrl += "-sharedstreams.icloud.com";
  baseUrl += "/";
  baseUrl += token;
  baseUrl += "/sharedstreams/";

  return baseUrl;
}

/**
 * Fetches the metadata for aple photos
 * @param {String} baseUrl Apple photos album url.
 */
function getPhotoMetadata(baseUrl) {
  var url = baseUrl + "webstream";

  var headers = {
    Origin: "https://www.icloud.com",
    "Accept-Language": "en-US,en;q=0.8",
    "User-Agent":
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36",
    "Content-Type": "text/plain",
    Accept: "*/*",
    Referer: "https://www.icloud.com/sharedalbum/",
    Connection: "keep-alive"
  };

  var dataString = '{"streamCtag":null}';

  var options = {
    url: url,
    method: "POST",
    headers: headers,
    body: dataString
  };

  return rp(options).then(function(body) {
    var data = JSON.parse(body);

    var photos = {};

    var photoGuids = [];

    data.photos.forEach(function(photo) {
      photos[photo.photoGuid] = photo;
      photoGuids.push(photo.photoGuid);
    });

    return {
      photos: photos,
      photoGuids: photoGuids
    };
  });
}

/**
 * Geg ULR for images
 * @param {String} baseUrl Directory to check for images as string.
 * @param {String} photoGuids
 */
function getUrls(baseUrl, photoGuids) {
  var url = baseUrl + "webasseturls";

  var headers = {
    Origin: "https://www.icloud.com",
    "Accept-Language": "en-US,en;q=0.8",
    "User-Agent":
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36",
    "Content-Type": "text/plain",
    Accept: "*/*",
    Referer: "https://www.icloud.com/sharedalbum/",
    Connection: "keep-alive"
  };

  var dataString = JSON.stringify({
    photoGuids: photoGuids
  });

  var options = {
    url: url,
    method: "POST",
    headers: headers,
    body: dataString
  };

  // console.log('Retrieving URLs for ' + photoGuids[0] + ' - ' + photoGuids[photoGuids.length - 1] + '...');

  return rp(options).then(function(body) {
    var data = JSON.parse(body);

    var items = {};

    for (var itemId in data.items) {
      var item = data.items[itemId];

      items[itemId] = "https://" + item.url_location + item.url_path;
    }

    return items;
  });
}

/**
 * Get best image from meta
 * @param {Array} metadata Directory to check for images as string.
 * @param {Array} urls
 */
function decorateUrls(metadata, urls) {
  for (var photoId in metadata.photos) {
    var photo = metadata.photos[photoId];

    var biggestFileSize = 0;
    var bestDerivative = null;

    for (var derivativeId in photo.derivatives) {
      var derivative = photo.derivatives[derivativeId];

      if (parseInt(derivative.fileSize, 10) > biggestFileSize) {
        biggestFileSize = parseInt(derivative.fileSize, 10);
        bestDerivative = derivative;
      }
    }

    if (bestDerivative) {
      if (typeof urls[bestDerivative.checksum] == "undefined") {
        continue;
      }

      var url = urls[bestDerivative.checksum];
      metadata.photos[photoId].url = url;
      metadata.photos[photoId].bestDerivative = bestDerivative;
    }
  }
}

/**
 * Download file to directory
 * @param {String} url File URL.
 * @param {String} dest Full path of destination folder to save file.
 * @param {String} cb Callback function. 
 */
function downloadFile(url, dest, cb) {
  const file = fs.createWriteStream(dest);
  const sendReq = request.get(url);

  // verify response code
  sendReq.on("response", response => {
    if (response.statusCode !== 200) {
      console.log("Response status was " + response.statusCode);
    }

    sendReq.pipe(file);
  });

  // close() is async, call cb after close completes
  file.on("finish", () => file.close(resizeFile(dest, maxWidth)));

  // check for request errors
  sendReq.on("error", err => {
    fs.unlink(dest);
    console.log(err.message);
  });

  file.on("error", err => {
    // Handle errors
    fs.unlink(dest); // Delete the file async. (But we don't check the result)
    console.log(err.message);
  });
}

/**
 * Resize image with specific width
 * @param {String} imageFile Full path of image file.
 * @param {Integer} maxWidth Max image width in pixels, default to 1000 px.
 */
function resizeFile(imageFile, maxWidth = 1000) {
  console.log("["+imageFile+"] - Analyzing");

  try {
    if (fs.existsSync(imageFile)) {

      // Resize image
      console.log("["+imageFile+"] - Resizing to "+ maxWidth + "px and compressing it to "+ compressQuality +"%.");
      sharp(imageFile)
        .resize(maxWidth,maxWidth, { fit: 'inside', withoutEnlargement: true })
        .jpeg({
          quality: compressQuality,
          chromaSubsampling: '4:4:4'
        })
        .toBuffer((err, buffer) => {
          if (err) {
            console.log(imageFile + " - " + err);
          }
          fs.writeFile(imageFile, buffer, (e) => {
            if(e){
              console.log("Error saving image to file: "+e);
            }
          });
        });
    }
  } catch (err) {
    console.error(err);
  }
}

/**
 * List images in directory 
 * @param {String} directory Directory to check for images as string.
 */
function listImagesInDir(directory) {
  var images = [];

  fs.readdir(directory, function(err, items) {
    extensions = ["jpg", "jpeg", "gif", "png", "bmp"];
    // Get image files
    for (var i = 0; i < items.length; i++) {
      if (extensions.includes(items[i].split(".").pop())) {
        images.push(items[i]);
      }
    }
  });

  return images;
}

/**
 * List images synchronously in directory
 * @param {String} directory Directory to check for images as string.
 */
function listImagesInDirSync (directory) {
  if (!fs.existsSync(directory)) return [];
  let newDirents = fs.readdirSync(directory, { withFileTypes: true });
  let files = newDirents
    .filter(d => !d.isDirectory() && d.isFile())                                      // remove directories
    .map(d => d.name)                                                                 // file name
    .filter(f => !(/(^|\/)\.[^\/\.]/g).test(f))                                       // remove files starting with . (macOS .DS_STORE for example)
    .filter(f => [".jpg", ".jpeg", ".gif", ".png", ".bmp"].includes(path.extname(f))) // remove files that are not an image
    .map(f => `${directory}/${f}`);                                                   // rebuild file path

  return files;
}

/**
 * Shuffles array in place.
 * @param {Array} a items An array containing the items.
 */
function shuffle(a) {
  var j, x, i;
  for (i = a.length - 1; i > 0; i--) {
      j = Math.floor(Math.random() * (i + 1));
      x = a[i];
      a[i] = a[j];
      a[j] = x;
  }
  return a;
}